Mathlib/SetTheory/Ordinal/Notation/FundamentalSequence.lean

```diff
@@ -0,0 +1,186 @@
+
+/-!
+# Fundamental sequences
+
+If `x` is a successor limit in a linear order `α`, a fundamental sequence for `x` is a strictly
+monotonic sequence `ℕ → α` whose supremum is `x`.
+
+If `x` is not a successor limit, we can define two other types of fundamental sequences: the empty
+sequence `∅` with limit `⊥`, and the singleton sequence `{x}` with limit `succ x` (the limit here is
+more properly the least strict upper bound). This ensures that every countable ordinal
+(nonconstructively) has a fundamental sequence, and it simplifies the definition of the fast-growing
+hierarchy.
+
+To nicely manage these three different cases, we define `Sequence α = Option α ⊕ (ℕ → α)` as the
+type of sequences with length 0, 1, or `ω`. At the moment, the file contains only the basic API for
+this type.
+
+The `Sequence` API is intended to be used constructively in the context of ordinal notations. For a
+more general but non-constructive notion of fundamental sequences for values with arbitrary
+cofinalities, see `Ordinal.IsFundamentalSequence`.
+
+## Todo
+
+- Define fundamental sequences
+- Define the slow and fast-growing hierarchies
+- Replace the definitions in `Mathlib.SetTheory.Ordinal.Notation.Cantor` by the ones in this file.
+-/
+
+universe u
+
+variable {α β : Type*}
+
+namespace Ordinal
+
+/-- The type of sequences with length 0, 1, or `ω`. These are the possible different lengths for a
+fundamental sequence. -/
+def Sequence (α : Type u) : Type u :=
+  Option α ⊕ (ℕ → α)
+
+namespace Sequence
+
+/-- The empty sequence, whose limit is the bottom element. -/
+instance : EmptyCollection (Sequence α) :=
+  ⟨Sum.inl none⟩
+
+instance : Inhabited (Sequence α) :=
+  ⟨∅⟩
+
+/-- The sequence consisting only of `x`, whose limit is the succesor of `x`. -/
+instance : Singleton α (Sequence α) :=
+  ⟨fun x ↦ Sum.inl (some x)⟩
+
+/-- A sequence `ℕ → α`, whose limit is its least strict upper bound, or equivalently for strictly
+monotonic sequences, its supremum. -/
+def ofFun (f : ℕ → α) : Sequence α :=
+  Sum.inr f
+
+@[simp]
+theorem singleton_ne_empty (x : α) : ({x} : Sequence α) ≠ ∅ := by
+  change Sum.inl _ ≠ Sum.inl _
+  simp
+
+@[simp] theorem ofFun_ne_empty (f : ℕ → α) : ofFun f ≠ ∅ := Sum.inr_ne_inl
+@[simp] theorem ofFun_ne_singleton (f : ℕ → α) (x : α) : ofFun f ≠ {x} := Sum.inr_ne_inl
+
+@[simp] theorem empty_ne_singleton (x : α) : ∅ ≠ ({x} : Sequence α) := (singleton_ne_empty x).symm
+@[simp] theorem empty_ne_ofFun (f : ℕ → α) : ∅ ≠ ofFun f := Sum.inl_ne_inr
+@[simp] theorem singleton_ne_ofFun (x : α) (f : ℕ → α) : {x} ≠ ofFun f := Sum.inl_ne_inr
+
+@[simp] theorem sum_inl_none_def : Sum.inl none = (∅ : Sequence α) := rfl
+@[simp] theorem sum_inl_some_def (x : α) : Sum.inl (some x) = ({x} : Sequence α) := rfl
+@[simp] theorem sum_inr_def (f : ℕ → α) : Sum.inr f = ofFun f := rfl
+
+/-- Recursion on sequences, using the preferred forms of the constructors. -/
+def recOn {p : Sequence α → Sort*} (s : Sequence α) (empty : p ∅) (singleton : ∀ x, p {x})
```

Comment:
```suggestion
/-- Recursion on sequences, using the preferred forms of the constructors. -/
@[elab_as_elim]
def recOn {p : Sequence α → Sort*} (s : Sequence α) (empty : p ∅) (singleton : ∀ x, p {x})
```
I think. @eric-wieser can you confirm this is what's useful?

Reply:
This makes sense I think, though I'm not sure what this attribute actually does.